### PR TITLE
[native_toolchain_c] Default handling for PIC/PIE compiler flags

### PR DESCRIPTION
**Pull request by blaugold**
The rust `cc` crate provides [defaults for passing `-fPIC`](https://docs.rs/cc/latest/cc/struct.Build.html#method.pic) to the compiler that simplify configuration for most use cases. This PR adapts this behavior for `CBuilder`.

Closes https://github.com/dart-lang/native/issues/119

### CODE REVIEW

File: pkgs/native_toolchain_c/lib/src/cbuilder/cbuilder.dart [hunk 1]
```diff
@@ -57,18 +57,27 @@ class CBuilder implements Builder {
...
     this.dartBuildFiles = const ['build.dart'],
     @visibleForTesting this.installName,
+    this.pic = true,
   }) : _type = _CBuilderType.library;
 
   CBuilder.executable({
     required this.name,
     this.sources = const [],
     this.dartBuildFiles = const ['build.dart'],
+    this.pic = false,
```
Comment (dcharkes) [inline pkgs/native_toolchain_c/lib/src/cbuilder/cbuilder.dart:80]: Maybe make this a named argument `pie` and assign to `this.pic` in the initializer list.

File: pkgs/native_toolchain_c/lib/src/cbuilder/cbuilder.dart [hunk 2]
```diff
@@ -57,18 +57,27 @@ class CBuilder implements Builder {
   @visibleForTesting
   final Uri? installName;
 
+  /// Whether the compiler will emit position independent code.
+  ///
+  /// When set to `null`, the default behavior of the compiler will be used.
+  ///
+  /// Defaults to `true` for libraries and `false` for executables.
+  final bool? pic;
```
Comment (dcharkes) [inline pkgs/native_toolchain_c/lib/src/cbuilder/cbuilder.dart:65]: Add a comment that this means PIE for executables

File: pkgs/native_toolchain_c/test/cbuilder/cbuilder_test.dart [hunk 3]
```diff
@@ -19,112 +19,116 @@ import 'package:test/test.dart';
...
-      );
-    });
+          expect(result.stdout.trim(), endsWith('Hello world.'));
+        });
+      });
+    }
+
+    for (final dryRun in [true, false]) {
+      final testSuffix = dryRun ? ' dry_run' : '';
+      test('Cbuilder dylib$testSuffix (pic: $pic)', () async {
```
Comment (dcharkes) [inline pkgs/native_toolchain_c/test/cbuilder/cbuilder_test.dart:77]: nit: please `+= testSuffix` instead.

File: pkgs/native_toolchain_c/test/cbuilder/cbuilder_test.dart [hunk 4]
```diff
@@ -19,112 +19,116 @@ import 'package:test/test.dart';
...
-      await inTempDir(
-        // https://github.com/dart-lang/sdk/issues/40159
-        keepTemp: Platform.isWindows,
-        (tempUri) async {
-          final addCUri =
-              packageUri.resolve('test/cbuilder/testfiles/add/src/add.c');
-          const name = 'add';
+  for (final pic in [null, true, false]) {
+    for (final buildMode in BuildMode.values) {
+      test('Cbuilder executable $buildMode (pic: $pic)', () async {
```
Comment (dcharkes) [inline pkgs/native_toolchain_c/test/cbuilder/cbuilder_test.dart:24]: Please make a testSuffix here with buildMode and pic, that scales better if/when we add even more surrounding for loops.

**Review by dcharkes — APPROVED** [review]
(I know it's still a draft.)

But LGTM.

Please add some more documentation to the PR description that refers to the CC crate using this as default.

**Comment by dcharkes** [comment]
Oh and ditto here for version number and changelog entry.

File: pkgs/native_toolchain_c/lib/src/cbuilder/cbuilder.dart (hunk 2 above)
Comment (blaugold) [reply pkgs/native_toolchain_c/lib/src/cbuilder/cbuilder.dart:96]: Done

File: pkgs/native_toolchain_c/lib/src/cbuilder/cbuilder.dart (hunk 1 above)
Comment (blaugold) [reply pkgs/native_toolchain_c/lib/src/cbuilder/cbuilder.dart]: Done

File: pkgs/native_toolchain_c/test/cbuilder/cbuilder_test.dart (hunk 4 above)
Comment (blaugold) [reply pkgs/native_toolchain_c/test/cbuilder/cbuilder_test.dart]: Done

File: pkgs/native_toolchain_c/test/cbuilder/cbuilder_test.dart (hunk 3 above)
Comment (blaugold) [reply pkgs/native_toolchain_c/test/cbuilder/cbuilder_test.dart]: Done

File: pkgs/native_toolchain_c/CHANGELOG.md [hunk 5]
```diff
@@ -3,6 +3,8 @@
 - Added `defines` for specifying custom defines.
 - Added `buildModeDefine` to toggle define for current build mode.
 - Added `ndebugDefine` to toggle define of `NDEBUG` for non-debug builds.
+- Generate position independent code for libraries by default and add
```
Comment (dcharkes) [inline pkgs/native_toolchain_c/CHANGELOG.md:6]: Auto-publishing expects us to publish the first commit with a non `-wip` suffix. So this has to go in 0.2.2.

**Review by blaugold — COMMENTED** [review]
<stroke>I've updated the changelog, but left the version, since 0.2.1 has not been published yet.</stroke>

File: pkgs/native_toolchain_c/CHANGELOG.md (hunk 5 above)
Comment (blaugold) [reply pkgs/native_toolchain_c/CHANGELOG.md:6]: Ah, sorry, when I looked this morning 0.2.1 was not published yet.

File: pkgs/native_toolchain_c/test/helpers.dart [hunk 6]
```diff
@@ -11,6 +11,32 @@ import 'package:native_toolchain_c/src/native_toolchain/apple_clang.dart';
...
+/// ```dart
+/// void main() {
+///   for (final dryRun in [true, false]) {
+///     final suffix = testSuffix([if (dryRun) 'dry_run']);
+///
+///     test('My test$suffix', () {});
+///   }
+/// }
+/// ```
+String testSuffix(List<Object> tags) => switch (tags) {
```
Comment (dcharkes) [inline pkgs/native_toolchain_c/test/helpers.dart:35]: Nice!

**Comment by dcharkes** [comment]
Thanks @blaugold! 🥇

**Comment by dcharkes** [comment]
cc @devoncarew I think we used to get code-coverage and publisher comments on PRs, but they are not on this PR. Did we change some settings? (The publish comment that has a clickable draft-release link is quite handy!)

**Pull request merged**